appliance/mariadb/process.go

```diff
@@ -0,0 +1,1106 @@
+
+	// Assert that the process is not running. This should not occur.
+	if p.running() {
+		panic(fmt.Sprintf("unexpected state running role=%s", p.config().Role))
+	}
+
+	if err := p.writeConfig(configData{ReadOnly: downstream != nil}); err != nil {
+		logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
+		return err
+	}
+
+	if err := p.installDB(); err != nil {
+		return err
+	}
+
+	if err := p.start(); err != nil {
+		return err
+	}
+
+	if err := p.initPrimaryDB(); err != nil {
+		if e := p.stop(); err != nil {
+			logger.Debug("ignoring error stopping process", "err", e)
+		}
+		return err
+	}
+
+	if downstream != nil {
+		p.waitForSync(downstream, true)
+	}
+
+	return nil
+}
+
+// Backup returns a reader for streaming a backup in xbstream format.
+func (p *Process) Backup() (io.ReadCloser, error) {
+	cmd := exec.Command(
+		filepath.Join(p.BinDir, "innobackupex"),
+		"--defaults-file="+p.ConfigPath(),
+		"--host=127.0.0.1",
+		"--port="+p.Port,
+		"--user=flynn",
+		"--password="+p.Password,
+		"--socket=",
+		"--stream=xbstream",
+		".",
+	)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		stdout.Close()
+		return nil, err
+	}
+	return stdout, nil
+}
+
+// Restore restores the database from an xbstream backup.
+func (p *Process) Restore(r io.ReadCloser) error {
+	if err := p.writeConfig(configData{}); err != nil {
+		return err
+	}
+	if err := p.unpackXbstream(r); err != nil {
+		return err
+	}
+	if err := p.restoreApplyLog(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Process) unpackXbstream(r io.ReadCloser) error {
+	defer r.Close()
+
+	cmd := exec.Command(filepath.Join(p.BinDir, "xbstream"), "-x", "--directory="+p.DataDir)
+	cmd.Stdin = r
+
+	if buf, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("xbstream failed", "err", err, "output", string(buf))
+		return err
+	}
+	return nil
+}
+
+func (p *Process) restoreApplyLog() error {
+	cmd := exec.Command(
+		filepath.Join(p.BinDir, "innobackupex"),
+		"--defaults-file="+p.ConfigPath(),
+		"--apply-log",
+		p.DataDir,
+	)
+	if buf, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("innobackupex apply-log failed", "err", err, "output", string(buf))
+		return err
+	}
+	return nil
+}
+
+func (p *Process) assumeStandby(upstream, downstream *discoverd.Instance) error {
+	logger := p.Logger.New("fn", "assumeStandby", "upstream", upstream.Addr)
+	logger.Info("starting up as standby")
+
+	if err := p.writeConfig(configData{ReadOnly: true}); err != nil {
+		logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
+		return err
+	}
+
+	if p.running() {
+		if err := p.stop(); err != nil {
+			return err
+		}
+	} else {
+		if p.WaitUpstream {
+			if err := p.waitForUpstream(upstream); err != nil {
+				return err
+			}
+		}
+
+		if err := func() error {
+			// TODO(benbjohnson): Lock table during backup: "FLUSH TABLES WITH READ LOCK"
```

Comment:
This isn't necessary. The master doesn't need to be locked when using `xtrabackup`, it will do a first pass copying physical files and keeping track of what happens in the log, then lock the tables and copy anything that changed since the backup started and also write the log and information about it into the backup stream.